config/prowgen/main.go

```diff
@@ -0,0 +1,183 @@
+generated output will be written to stdout.
+`
+)
+
+var (
+	generateProwExample = fmt.Sprintf(`
+To generate tests for the a branch called foo:
+
+	%s --branch=foo
+`, generateProwCommand)
+)
+
+type generateProwOptions struct {
+	// Branch specifies the name of the branch whose tests should be generated
+	Branch string
+
+	// OutputDir specifies the dir to output the yaml files to. If empty, output
+	// will be written to stdout.
+	OutputDir string
+}
+
+func (o *generateProwOptions) AddFlags(fs *flag.FlagSet, markRequired func(string)) {
+	fs.StringVar(&o.Branch, "branch", "", fmt.Sprintf("Type of tests to generate; one of ('*' generates all branches) %v", append(prowspecs.KnownBranches(), "*")))
+	fs.StringVarP(&o.OutputDir, "output-dir", "o", "", "OutputDir specifies the dir to output the yaml files to. If empty, output will be written to stdout.")
+
+	markRequired("branch")
+}
+
+func generateProwCmd() *cobra.Command {
+	o := &generateProwOptions{}
+
+	cmd := &cobra.Command{
+		Use:          generateProwCommand,
+		Short:        generateProwDescription,
+		Long:         generateProwLongDescription,
+		Example:      generateProwExample,
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if o.Branch == "*" {
+				for _, branch := range prowspecs.KnownBranches() {
+					if err := o.runGenerateProw(branch); err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+			return o.runGenerateProw(o.Branch)
+		},
+	}
+
+	o.AddFlags(cmd.Flags(), func(s string) {
+		if err := cmd.MarkFlagRequired(s); err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	})
+
+	return cmd
+}
+
+// sanitizedArgs strips the path from the command which was used to invoke the script,
+// so we don't include things like "/home/workspace/release/bin/prowgen"
+func sanitizedArgs() []string {
+	args := os.Args[:]
+
+	for i := range args {
+		if !strings.Contains(args[i], "/") {
+			continue
+		}
+
+		args[i] = filepath.Base(args[i])
+	}
+
+	return args
+}
+
+func (o *generateProwOptions) runGenerateProw(branch string) error {
+	spec, err := prowspecs.SpecForBranch(branch)
+	if err != nil {
+		return err
+	}
+
+	jobFile := spec.GenerateJobFile()
+
+	out, err := yaml.Marshal(jobFile)
+	if err != nil {
+		return err
+	}
+
+	prelude := fmt.Sprintf(
+		`# THIS FILE HAS BEEN AUTOMATICALLY GENERATED
+# Don't manually edit it; instead edit the "prowgen" tool which generated it
+# Generated with: %s
+
+`,
+		strings.Join(sanitizedArgs(), " "),
+	)
+
+	data := prelude + string(out)
+
+	if o.OutputDir == "" {
+		fmt.Println(data)
+		return nil
+	}
+
+	branchPath := path.Join(o.OutputDir, branch)
+
+	if err := os.MkdirAll(branchPath, 0755); err != nil {
+		return err
+	}
+
+	path := filepath.Join(branchPath, fmt.Sprintf("cert-manager-%s.yaml", branch))
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(f, strings.NewReader(data)); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
I prefer to be able to output to stdout, inspect it and then, if I want to, redirect it to a file when I am making some changes by hand rather than having to figure out how to pass a path to a file and read it from there. I think we should leave the stdout option in - any templating tool I can think of can write to stdout.

Could we perhaps only make changes that are required to move these files in this PR and not additional unrelated changes?

Reply:
on line 152 we still have `if o.OutputDir == "" {` which allows us to output to console.
The reason I did change this logic is that I wanted to be able to specify where to output the files (if not via stdout).

Reply:
@irbekrm I'll create a seperate PR that just copies the code, so this PR becomes easier to review.

Reply:
@irbekrm  https://github.com/jetstack/testing/pull/880